### ISSUE TITLE
Unable to add multiple `--add-opens` args to forked Groovy compiler

### ISSUE THREAD

**Issue by bigdaz** [gradle/gradle#7045]
Because of the `Set`s used when [merging fork options](https://github.com/gradle/gradle/blob/3ccee91ec332e3ebde5a6342f03b9480f2581b2a/subprojects/language-jvm/src/main/java/org/gradle/api/internal/tasks/compile/daemon/AbstractDaemonCompiler.java#L66-L74), any duplicate jvm-args will be removed. This makes it impossible to supply multiple `--add-opens` parameters to the JVM.
For example:
```
compileTestGroovy {
    options.forkOptions.jvmArgs += ['--add-opens', 'java.base/java.lang=ALL-UNNAMED']
    options.forkOptions.jvmArgs += ['--add-opens', 'java.base/java.lang.invoke=ALL-UNNAMED']
}
```

Results in a command like like `java --add-opens java.base/java.lang=ALL-UNNAMED java.base/java.lang.invoke=ALL-UNNAMED ...` and a failure like https://e.grdev.net/s/ydiasku6wfnji/console-log.

**Comment by baron1405** [comment]
Hey @bigdaz, I think I am seeing this problem when IntelliJ tries to start the Gradle daemon. I load a project into IntelliJ using the Tooling API and see the following out:
```
...
> Configure project :

BUILDING gradle-core-plugins - Gradle: 5.0, Kotlin: 1.3.10, Java: 11.0.1


Deprecated Gradle features were used in this build, making it incompatible with Gradle 6.0.
Use '--warning-mode all' to show the individual deprecation warnings.
See https://docs.gradle.org/5.0/userguide/command_line_interface.html#sec:command_line_warnings

CONFIGURE SUCCESSFUL in 32s
Unable to start the daemon process.
This problem might be caused by incorrect configuration of the daemon.
For example, an unrecognized jvm option is used.
Please refer to the user guide chapter on the daemon at https://docs.gradle.org/5.0/userguide/gradle_daemon.html
Please read the following process output to find out more:
-----------------------
Error: Could not find or load main class java.base.java.lang=ALL-UNNAMED
Caused by: java.lang.ClassNotFoundException: java.base.java.lang=ALL-UNNAMED
```
Looks like the second `--add-options` is be added to the first and we end up with an invalid java command line where the second value is being mistreated as the main class.

My environment is:
* Gradle 5.0
* IntelliJ 2018.3.1
* OpenJDK 11.0.1
* CentOS 7.6
* Verified that IntelliJ is using the same JDK when running Gradle as the command-line

The problem is repeatable both when loading the project or refreshing it using the Gradle tool view in IntelliJ. Note that building the project on the command-line does not have this problem.

**Comment by baron1405** [comment]
Did some hunting on the JetBrains issue tracker and found 2 issues identical to mine. They are:

* https://youtrack.jetbrains.com/issue/IDEA-202975
* https://youtrack.jetbrains.com/issue/IDEA-203429

There is clearly a problem but the amount of fumbling around with JDK versions in various places makes it hard to follow whether anyone has found a viable workaround other than forcing JDK 8.

**Comment by baron1405** [comment]
@bigdaz do you have any suggestions for how I can set debug logging for the Gradle deamon launcher that is called from within IntelliJ? I tried setting org.gradle.logging.level in my project's gradle.properties but no luck. I think IntelliJ is running calling the tooling API to start the deamon but I guess it is not going through a path that picks up my project's gradle.properties.

**Comment by Xerus2000** [comment]
Indeed, this issue breaks using IDEA with Gradle 5.0 for me. What is going on here?

**Comment by bigdaz** [comment]
@baron1405 @Xerus2000 IntelliJ import works fine for the `gradle` build (and many others) with JDK 11, so there must be something specific to your projects that is causing this issue.

Any chance you can put together a self-contained project that would allow us to reproduce this issue?Failing that, a full stack-trace might help. 

Note that the this issue is about failure to launch a forked Groovy compiler: it's not clear from your report that this is the same issue, although the symptoms are very similar.

**Comment by baron1405** [comment]
@bigdaz how do we get debug output from the Gradle daemon run from within IntelliJ via the Tooling API?

**Comment by baron1405** [comment]
@bigdaz I think I have found the problem. The bad news is that any Gradle build using JDK 11 and containing a buildSrc will hit this issue. The manifestation of the problem matches the Groovy problem because it is due to Groovy.

IntelliJ configures a Gradle project through the Gradle Tooling API using a command line such as (excerpt from IntelliJ log):
```
2019-01-19 00:56:57,853 [1761362]   INFO - s.plugins.gradle.GradleManager - Instructing gradle to use java from /opt/java/11 
2019-01-19 00:56:57,853 [1761362]   INFO - s.plugins.gradle.GradleManager - Instructing gradle to use java from /opt/java/11 
2019-01-19 00:56:57,864 [1761373]   INFO - xecution.GradleExecutionHelper - Passing command-line args to Gradle Tooling API: -Didea.version=2018.3.3 -Djava.awt.headless=true -Pandroid.injected.build.model.only=true -Pandroid.injected.build.model.only.advanced=true -Pandroid.injected.invoked.from.ide=true -Pandroid.injected.build.model.only.versioned=3 -Pandroid.builder.sdkDownload=false --init-script /tmp/ijinit2.gradle 
2019-01-19 00:56:58,862 [1762371]   INFO - xecution.GradleExecutionHelper - Passing command-line args to Gradle Tooling API: -Didea.version=2018.3.3 -Djava.awt.headless=true -Pandroid.injected.build.model.only=true -Pandroid.injected.build.model.only.advanced=true -Pandroid.injected.invoked.from.ide=true -Pandroid.injected.build.model.only.versioned=3 -Pandroid.builder.sdkDownload=false --init-script /tmp/ijinit2.gradle -Didea.version=2018.3.3 -Djava.awt.headless=true -Pandroid.injected.build.model.only=true -Pandroid.injected.build.model.only.advanced=true -Pandroid.injected.invoked.from.ide=true -Pandroid.injected.build.model.only.versioned=3 -Pandroid.builder.sdkDownload=false --init-script /tmp/ijinit3.gradle 
```
Notice the last part of the parameters is an init script (/tmp/ijinit3.gradle). Looking at that script, I found the following:
```
allprojects { project ->
  if (project.name == 'buildSrc') {
    apply plugin: 'groovy'
    dependencies {
      compile gradleApi()
      compile localGroovy()
    }
  }
}
```
In other words, IntelliJ is always applying the Groovy plugin and adding Groovy to a project that has a buildSrc. So any problems that Groovy has with Java 11 will show up when IntelliJ attempts to configure a Gradle project suing Java 11 and containing a buildSrc.

I may be missing something but the pieces seem to fit together. I have not yet explored how to work around the problem but wanted to post my findings as soon as possible.

**Comment by baron1405** [comment]
After spending a few hours trying various ideas for working around this issue, I have concluded that until the Groovy options merge bug is fixed in Gradle, I cannot use JDK 11 with my project that has a buildSrc and instead must revert to JDK 8. @bigdaz any status on getting a fix into 5.2?

**Comment by baron1405** [comment]
@Xerus2000 OK but then you are using Groovy in the project?

**Comment by bigdaz** [comment]
@baron1405 Thanks for all of your analysis on this issue. I'm not actively working a fix, and it's unlikely that a fix will be in 5.2.

The gradle build itself uses a combination of JDK11 together with `buildSrc` containing both Groovy and Kotlin code. It imports into IntelliJ just fine, so there must be something specific to your setup that's causing this. While the symptoms are the same, I'm not certain that fixing this bug will fix your issue.

Here are a couple of things you could do to help progress this issue:
1. Take your problematic build, and simplify it until you have a minimal example build that exhibits this issue. Sharing a project like this makes it a lot easier to reproduce the issue, and much easier to fix.
2. Try patching Gradle yourself so that duplicate JVM args are not removed [here](https://github.com/gradle/gradle/blob/3ccee91ec332e3ebde5a6342f03b9480f2581b2a/subprojects/language-jvm/src/main/java/org/gradle/api/internal/tasks/compile/daemon/AbstractDaemonCompiler.java#L70-L71). This could be as simple as using a `List` instead of a `Set`, but that depends how the compiler deals with duplicate arguments.

**Comment by baron1405** [comment]
@bigdaz Thanks for the update! One question, does the Gradle build project perform the JVM args merging?

**Comment by bigdaz** [comment]
> One question, does the Gradle build project perform the JVM args merging?

Nope, the merging is done by the Groovy plugin, only when executed with "forked" mode. This causes the compiler to load in a separate JVM process, and it's this process that is getting the bad JVM args.

See https://docs.gradle.org/current/dsl/org.gradle.api.tasks.compile.GroovyCompileOptions.html#org.gradle.api.tasks.compile.GroovyCompileOptions:fork

**Comment by baron1405** [comment]
Perhaps that is the missing difference. My build specifies an org.gradle.java.home in its gradle.properties. That would cause the forking and lead to the Groovy plugin merging.

**Comment by Xerus2000** [comment]
ah, now we're getting somewhere. In my project I also specify a gradle java home!

**Comment by bigdaz** [comment]
@baron1405 I looked into this a bit more, and I am unable to reproduce the issue you're seeing.

Here's what I did (using Gradle 5.1.1):
- Create a new project using `gradle init` (accepting all defaults)
- Create a `gradle.properties` file in the project root directory specifying `org.gradle.java.home=<my-oracle-jdk-9.0.4>`
- Build directly with Gradle. 
    -Build succeeds.
- Import into IDEA using defaults, specifying JDK 11 as the project JDK.
    - Import and sync succeed.

Can you please try something similar, so we can work out the true conditions that cause this failure?

**Comment by baron1405** [comment]
@bigdaz I will try to repro with a simple project and report back. In the meantime, can you try adding the following to your build.gradle:
```
    withType<Test>().configureEach {
        // Silence the warnings about reflective access by explicitly allowing it
        jvmArgs("--illegal-access=deny")
        jvmArgs("--add-opens", "java.base/java.util=ALL-UNNAMED")
        jvmArgs("--add-opens", "java.base/java.lang=ALL-UNNAMED")
    }
```
Note that the above is in Kotlin so convert to Groovy as needed.

I had made the assumption that multiple `--add-opens` were being specified internally by Gradle to silence the reflective access noise. If only one is specified, the merge will produce a valid command-line. More than one `--add-opens` should cause the problem.

**Comment by bigdaz** [comment]
Adding that configuration to the build script did not impact execution in Gradle, or importing into IDEA.

**Comment by Xerus2000** [comment]
Is the `org.gradle.java.home` you specified different from the default JDK on your system? Could it be related to the OS? I am using KDE Neon, an Ubuntu-derivative.

**Comment by baron1405** [comment]
@Xerus2000 the JDK specified in `org.gradle.java.home` that leads to the problem is JDK 11 and is *not* the default on the system. The default on the system is JDK 8. In my case the OS CentOS 7.6.

**Comment by Xerus2000** [comment]
> @Xerus2000 the JDK specified in `org.gradle.java.home` that leads to the problem is JDK 11 and is _not_ the default on the system. The default on the system is JDK 8.

Yes, exactly the same situation.

**Comment by baron1405** [comment]
@bigdaz I have reproduced the problem in a simple project. See the [gradle7045 project](https://github.com/baron1405/gradle7045) and follow the instructions in the README to reproduce the problem. Everything was fine until I added the buildSrc project.

**Comment by bigdaz** [comment]
That's great, thanks @baron1405.  I've raised #8353 for this specific issue, which I believe is different from this one, which is specific to the Groovy compiler daemon.
